src/main/java/com/iexec/worker/executor/TaskExecutorService.java

```diff
@@ -30,122 +47,208 @@
+                .thenApply(stdout -> resultService.saveResult(chainTaskId, replicateModel, stdout)) // save result
+                .thenAccept(isSaved -> {if (isSaved) contribute(contributionAuth);})                // contribute
+                .handle((res, err) -> {                                                             // handle errors
+                    if (err != null) {
+                        log.error(err.getMessage());
+                        return null;
+                    }
+                    return res;
                 });
     }
 
-    private boolean executeTask(AvailableReplicateModel replicateModel) {
-        String chainTaskId = replicateModel.getContributionAuthorization().getChainTaskId();
+    @Async
+    private String compute(String chainTaskId, AvailableReplicateModel replicateModel) {
         String stdout = "";
 
-        if (contributionService.isChainTaskInitialized(chainTaskId)) {
-            feignClient.updateReplicateStatus(chainTaskId, RUNNING);
-
-            if (replicateModel.getAppType().equals(DappType.DOCKER)) {
-                //App
-                feignClient.updateReplicateStatus(chainTaskId, APP_DOWNLOADING);
-                boolean isAppDownloaded = dockerComputationService.dockerPull(chainTaskId, replicateModel.getAppUri());
-                if (isAppDownloaded) {
-                    feignClient.updateReplicateStatus(chainTaskId, APP_DOWNLOADED);
-                } else {
-                    stdout = "Failed to pull image";
-                    feignClient.updateReplicateStatus(chainTaskId, APP_DOWNLOAD_FAILED);
-                }
-                //Data
-                feignClient.updateReplicateStatus(chainTaskId, DATA_DOWNLOADING);
-                boolean isDatasetDownloaded = datasetService.downloadDataset(chainTaskId, replicateModel.getDatasetUri());
-                if (isDatasetDownloaded) {
-                    feignClient.updateReplicateStatus(chainTaskId, DATA_DOWNLOADED);
-                } else {
-                    stdout = "Failed to pull dataset";
-                    feignClient.updateReplicateStatus(chainTaskId, DATA_DOWNLOAD_FAILED);
-                }
-                //Compute
-                feignClient.updateReplicateStatus(chainTaskId, COMPUTING);
-                if (isAppDownloaded && isDatasetDownloaded) {
-                    stdout = dockerComputationService.dockerRunAndGetLogs(replicateModel);
-                }
-                if (!stdout.isEmpty()) {
-                    feignClient.updateReplicateStatus(chainTaskId, COMPUTED);
-                } else {
-                    stdout = "Failed to start computation";
-                    feignClient.updateReplicateStatus(chainTaskId, COMPUTE_FAILED);
-                }
-
-                resultService.saveResultInfo(chainTaskId, replicateModel, stdout);
-            }
-        } else {
-            log.warn("Task NOT initialized on chain [chainTaskId:{}]", chainTaskId);
+        // check app type
```

Comment:
!! Careful here! We don't check anymore that the task is initialized on-chain. We must do it before starting any anything.

Reply:
isn't it the role of the core ? not to send tasks unless they are initialized !

Reply:
I'll do it anyway, I think you're right

Reply:
When it is blockchain related, the worker always check before doing anything.

Reply:
👍